Avoid crash when module diff has non utf8 translatable text

Yannig opened the pull request:
Try to call to_text against result. If we get an exception, we return the text as bytes.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME

Crash when using diff option with iso8859-15 files.

##### ANSIBLE VERSION

```
ansible 2.3.0 (diff-iso8859-15 2d2d12f0cc) last updated 2017/02/22 23:45:17 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY

Fixes #21803. Try to call to_text against result string. If we get an exception, we return text as bytes.

```
ansible-playbook crash-diff.yml --diff
PLAY [Crash when using --diff option] ***********************************************************************************************************

TASK [Copy a dummy iso8859-15 file] *************************************************************************************************************
--- before: /tmp/test.txt
+++ after: /home/drayan/dev/yannig-ansible-playbooks/diff-iso8859-15/iso8859-15.txt
@@ -1,3 +1,3 @@
-Un super fichier avec é en iso8859-15
+Un super fichier avec � en iso8859-15
 
-A super file with é in iso8859-15
+A super file with � in iso8859-15

changed: [localhost]
```

Yannig committed:
Fix for #21803.
Try to call to_text against result. If we get an exception, we return the text as bytes.

abadger commented on lib/ansible/executor/task_executor.py:
```diff
@@ -134,7 +134,10 @@ def _clean_res(res):
+                    try:
+                        return to_text(res, errors='surrogate_or_strict')
+                    except:
+                        return bytes(res)
```
I think this will just lead to more errors later (as we're now returning either text or bytes rather than always returning text.

@jimi-c is it okay to replace bytes with unicode replacement characters here?  If wo, we can just use to_text's default error handler.

Yannig replied:
@abadger maybe I can try to remove bytes with there respective hex representation like this:

```
Un super fichier avec \xe9 en iso8859-15

A super file with \xe9 in iso8859-15
```

What do you think about?

Yannig replied:
Here's a code snippet:

```python
to_text(''.join([i if ord(i) < 128 else ("\\x%x" % ord(i)) for i in res]), errors='surrogate_or_strict')
```

jimi-c replied:
I have concerns that that will be extraordinarily slow with large results.

Yannig replied:
In this case, we can add a simple test on the diff size. If over a given size (maybe 1M?), we don't send the diff back showing only a message.

abadger replied:
@Yannig @jimi-c also thinks that  doing this generally will impact matching on the results (not just the diff output results).  He suggests catching the unicodeerror and raising a nicer error about only being able to use diff with utf8 file contents.

If we're able to tell that we're using diff here, though, maybe we could run to_text([..], error='surrogate_or_replace') just for the diff entry?  @jimi-c?

abadger commented:
I've pinged jimi-c on slack as well.  This will need revision afterwards, though.

needs_revision

Yannig committed:
Taking into account feedback:
- Intercept only UnicodeDecodeError
- Transform characters over 128 into there hex representation and send back utf8 string.

bcoca commented:
I would add warning to the exception 'We were unable to decode all characters, skipped some in effort to return as much as possible'

Yannig commented:
@bcoca done.

abadger commented:
@jimi-c, @Yannig does this satisfy both of your concerns? https://gist.github.com/abadger/7098ef7e89ef2b7d00d7193f31d83ea6

abadger commented:
@Yannig Cool, I'll PR that solution then.

Yannig commented:
Nice :)

bcoca commented on lib/ansible/executor/task_executor.py:
```diff
@@ -134,7 +134,12 @@ def _clean_res(res):
+                    try:
+                        return to_text(res, errors='surrogate_or_strict')
+                    except UnicodeDecodeError:
+                        display.warning("We were unable to decode all characters, skipped some in effort to return as much as possible")
```
we might want to be more specific 'all characters returned by the module'
